### PR TITLE
[WIP] Plugin Ecosystem

### PR DESCRIPTION
**Pull request by ConnorVG**
Eventually, this will solve #1507 (if accepted).

A lot is yet to be done such as writing tests, documentation, adding missing exceptions, handling the exceptions etc. This is in a very **pre** thought.

Notes will come when I can add them, will most likely be after I implement 'disabling' / 'uninstalling' of plugins. Currently this only installs and enables them. A demo plugin can be found at [ConnorVG/cachet-demo-plugin](https://github.com/connorvg/cachet-demo-plugin).

### CODE REVIEW

File: composer.lock [hunk 1]
```diff
@@ -4398,6 +4398,49 @@
             "time": "2016-05-06 18:25:35"
         },
         {
+            "name": "frlnc/php-slack",
```
Comment (jbrooksuk) [inline composer.lock]: Whoops, don't commit this file.

Comment (ConnorVG) [inline composer.lock]: Whoops :laughing:

File: app/Composers/DashboardComposer.php [hunk 2]
```diff
@@ -32,5 +33,91 @@ public function compose(View $view)
         $view->withIncidentTemplateCount(IncidentTemplate::count());
         $view->withComponentCount(Component::all()->count());
         $view->withSubscriberCount(Subscriber::isVerified()->count());
+        $view->withPluginCount(Plugin::count());
```
Comment (jbrooksuk) [inline app/Composers/DashboardComposer.php]: Should show the active plugin count.

**Comment by jbrooksuk** [comment]
@ConnorVG can you undo the metric stuff, please? :)

My plan is to move them into a Plugin once we know the system works and that we have a good way of handling everything. Thanks!

**Comment by ConnorVG** [comment]
I will do after I've confirmed the abilities of the plugin eco system with connorvg/cachet-metrics-plugin

On 14 Aug 2016 19:13, James Brooks <notifications@github.com> wrote:

@ConnorVG<https://github.com/ConnorVG> can you undo the metric stuff, please? :)

My plan is to move them into a Plugin once we know the system works and that we have a good way of handling everything. Thanks!

**Comment by jbrooksuk** [comment]
Awesome, thanks!

**Comment by jbrooksuk** [comment]
Dude, can you not drop the commit on a rebase to put metrics back?

**Comment by ConnorVG** [comment]
I actually want to keep the two commits, they have some other stuff in them for fixing reverse-testing (Plugin -> Cachet).

**Comment by ConnorVG** [comment]
We will drop all commits at the end, most likely in a new PR. So this is just one nice commitski-s.

**Comment by jbrooksuk** [comment]
Okay cool, just seems weird removing then reverting it in one commit.

File: app/Composers/ModulesComposer.php [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+     *
+     * @return void
+     */
+    public function compose(View $view)
+    {
+        $key = $view->getName();
+
+        $view->with('view', $key);
+
+        $modules = 'view.modules: '.$view->getName();
```
Comment (jbrooksuk) [inline app/Composers/ModulesComposer.php]: This can just be `$key`.

File: app/Composers/ModulesComposer.php [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        $groups = 'view.groups: '.$view->getName();
```
Comment (jbrooksuk) [inline app/Composers/ModulesComposer.php]: This too.

Comment (ConnorVG) [inline app/Composers/ModulesComposer.php]: :facepalm:

File: app/Composers/ModulesComposer.php (hunk 3 above)
Comment (ConnorVG) [inline app/Composers/ModulesComposer.php]: :facepalm:

**Comment by ConnorVG** [comment]
The tests will obviously fail due to the test existence tests. Tests clearly need writing still.

File: app/Integrations/Packagist/Packages.php [hunk 5]
```diff
@@ -171,7 +171,7 @@ public function download(array $package)
         $meta = stream_get_meta_data($temp);
         $path = $meta['uri'];
 
-        ($zip = new ZipArchive())->open($path);
+        with($zip = new ZipArchive())->open($path);
```
Comment (GrahamCampbell) [inline app/Integrations/Packagist/Packages.php]: we don't need this?

Comment (ConnorVG) [inline app/Integrations/Packagist/Packages.php]: For the 5.6 version we're testing with on Travis, we do.

https://travis-ci.org/CachetHQ/Cachet/jobs/153497870

Comment (GrahamCampbell) [inline app/Integrations/Packagist/Packages.php]: Hmmm. Must be due to the variable assignment that php 5.5 doesn't support this. Normally it does support this syntax.

Comment (ConnorVG) [inline app/Integrations/Packagist/Packages.php]: Yeah, I'm bemused at this one. _iirc_ all 5.6 should support this syntax.

Comment (ConnorVG) [inline app/Integrations/Packagist/Packages.php]: Variable instantiation it is. Wasn't added until _after_ PHP 5.6.

>Output for 5.6.0 - 5.6.25

```bash
Parse error: syntax error, unexpected '->' (T_OBJECT_OPERATOR) in /in/pNWt0 on line 5

Process exited with code 255.
```

Comment (GrahamCampbell) [inline app/Integrations/Packagist/Packages.php]: 😭

Comment (ConnorVG) [inline app/Integrations/Packagist/Packages.php]: It makes sense when you think about it tho. The lexer in <7.0 is shit (in comparison to 7.0, that is).

Comment (GrahamCampbell) [inline app/Integrations/Packagist/Packages.php]: brutal :trollface:

File: app/Bus/Events/Plugin/PluginWillBeInstalledEvent.php [hunk 6]
```diff
@@ -34,7 +34,7 @@
      *
      * @return void
      */
-    public function __construct(string $name)
```
Comment (GrahamCampbell) [inline app/Bus/Events/Plugin/PluginWillBeInstalledEvent.php]: yolo

Comment (ConnorVG) [inline app/Bus/Events/Plugin/PluginWillBeInstalledEvent.php]: yolo indeed :trollface:

**Comment by jbrooksuk** [comment]
So many events!

**Comment by ConnorVG** [comment]
@jbrooksuk the entire app needs to be juiced with fifteen million different events for everything **imo**.

**Comment by jbrooksuk** [comment]
True, true.

**Comment by ConnorVG** [comment]
Even though I'm pretty much rewriting this, I'm going to be sad to see this branch go. :sob:

**Comment by ConnorVG** [comment]
😉

**Pull request closed without merging**